lib/puppet/agent.rb

```diff
@@ -1,6 +1,167 @@
 require 'puppet/application'
 require 'puppet/error'
 
+# A class providing pre/post fork callbacks modeled after those registered
+# by the `pthread_atfork` function.
+# The callbacks defined in this class are no-ops on all platforms except
+# Solaris where they ensure the forked process runs in a different contract
+# than the parent process. This is necessary in order for the child process
+# (an agent run) to be able to restart/stop the puppet agent service without
+# being killed (as a result of the termination of the parent process
+# contract during the service restart/stop).
+# The implementation uses Fiddle to make calls to functions defined in the
+# libcontract library.
+class AtFork
+  require 'fiddle'
+
+  raise LoadError, %q(The loaded fiddle version doesn't define 'Fiddle::Handle') unless defined?(Fiddle::Handle)
```

Comment:
A bit of explanation, which hopefully answers some of your questions:
* Without this check travis was failing on ruby 1.9.3 which apparently includes an early version of Fiddle which doesn't define the `Fiddle::Handle` class and does define (and throws) different exceptions (e.g. `DL::DLError` instead of `Fiddle::DLError`) than more modern versions, so in order to avoid having to complicate the implementation to make it compatible with this older versions of fiddle I opted for adding the check.
* The check is really meant to ensure the code is executing with a sufficiently modern version of Fiddle - ideally I'd do something like:
```ruby
raise "Too old version of fiddle." unless Fiddle::Version >= x.y.z
```
and as such it felt reasonable to place it at the beginning.
* I agree that checking for whether the code is executing on Solaris is a more robust approach than the no-op fallback.